File: android/src/main/kotlin/com/bolddesksupportsdk/bd_support_sdk/BolddeskSupportSdkPlugin.kt
package com.bolddesksupportsdk.bd_support_sdk

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.syncfusion.bolddeskmobileSDK.BoldDeskSupportSDK
import com.syncfusion.bolddeskmobileSDK.BoldDeskSDKHome
import com.syncfusion.bolddeskmobileSDK.R

/** BolddeskSupportSdkPlugin */
class BolddeskSupportSdkPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "bd_support_sdk")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "initialize") {
            val appId = call.argument<String>("appId") ?: ""
            val brandUrl = call.argument<String>("brandUrl") ?: ""
            val onSuccessCallback: ((String) -> Unit)? = { successMessage ->
               result.success(mapOf("success" to true, "message" to successMessage))
            }

            val onErrorCallback: ((String) -> Unit)? = { errorMessage ->
               result.success(mapOf("success" to false, "message" to errorMessage))
            }
            try {
                BoldDeskSupportSDK.initialize(context, appId, brandUrl, onSuccessCallback, onErrorCallback)
            } catch (e: Exception) {
                result.error("INITIALIZATION_FAILED", e.message, null)
            }
        } else if (call.method == "loginWithJWTToken") {
             val jwtToken = call.argument<String>("jwtToken") ?: ""
             val onSuccessCallback: ((String) -> Unit)? = { successMessage ->
               result.success(mapOf("success" to true, "message" to successMessage))
            }

            val onErrorCallback: ((String) -> Unit)? = { errorMessage ->
               result.success(mapOf("success" to false, "message" to errorMessage))
            }
            try {
                BoldDeskSupportSDK.loginWithJWTToken(context, jwtToken, onSuccessCallback, onErrorCallback)
            } catch (e: Exception) {
                result.error("LOGIN_FAILED", e.message, null)
            }
        } else if (call.method == "logout") {
            BoldDeskSupportSDK.logout(context)
            result.success("Logged out Successfully")
        } else if (call.method == "showHomeDashboard") {
            BoldDeskSupportSDK.showHomeDashboard(context)
            result.success("Opened HomePage Successfully")
        } else if (call.method == "showKB") {
            BoldDeskSupportSDK.showKB(context)
            result.success("Opened KnowledgeBase Successfully")
        } else if (call.method == "showCreateTicket") {
            BoldDeskSupportSDK.showCreateTicket(context)
            result.success("Opened create ticket page Successfully")
        } else if (call.method == "showNotification") {
            val messageData = call.argument<Map<String, String>>("body") ?: emptyMap()
            val icon = call.argument<String>("icon") ?: ""
            val iconResId = resolveIcon(context, icon)
            try {
                BoldDeskSupportSDK.handlePushNotifications(context, messageData, iconResId)
                result.success("Notification shown Successfully")
            } catch (e: Exception) {
                result.error("NOTIFICATION_FAILED", e.message, null)
            }
        } else if (call.method == "setFCMRegistrationToken") {
            val token = call.argument<String>("token") ?: ""
            BoldDeskSupportSDK.setFCMRegistrationToken(context,token)
            result.success("Token Set Successfully")
        } else if (call.method == "setLoggingEnabled"){
            val isEnabled = call.argument<Boolean>("enable") ?: false
            BoldDeskSupportSDK.setLoggingEnabled(isEnabled)
            result.success("Logging set Successfully")
        } else if (call.method == "applyTheme") {
            val primaryColor = call.argument<String>("primaryColor") ?: ""
            val accentColor = call.argument<String>("accentColor") ?: ""
            BoldDeskSupportSDK.applyTheme(primaryColor, accentColor)
            result.success("Theme applied Successfully")
        } else if (call.method == "setPreferredTheme") {
            val theme = call.argument<String>("theme") ?: ""
            BoldDeskSupportSDK.setPreferredTheme(theme)
            result.success("Theme applied Successfully")
        }else if (call.method == "setSystemFontSize"){
            val isEnabled = call.argument<Boolean>("enable") ?: false
            BoldDeskSupportSDK.setSystemFontSize(isEnabled)
            result.success("Logging set Successfully")
        }else if (call.method == "applyCustomFontFamilyInAndroid") {
            val regular = call.argument<String>("regular") ?: ""
            val medium = call.argument<String>("medium") ?: ""
            val semiBold = call.argument<String>("semiBold") ?: ""
            val bold = call.argument<String>("bold") ?: ""
            try {
                // Dynamically find the resource ID for each font name
                val regularId = getFontResourceId(context,regular)
                val mediumId = getFontResourceId(context,medium)
                val semiBoldId = getFontResourceId(context,semiBold)
                val boldId = getFontResourceId(context,bold)

                // Call YOUR Android SDK with the integer IDs you just found
                BoldDeskSupportSDK.applyCustomFontFamily(
                    regular = regularId,
                    medium = mediumId,
                    semiBold = semiBoldId,
                    bold = boldId
                )
                result.success(null)

            } catch (e: Exception) {
                result.error("FONT_NOT_FOUND", e.message, null)
            }
        } else if (call.method == "setHeaderLogo"){
            val logoUrl = call.argument<String>("logoURL") ?: ""
            BoldDeskSDKHome.setHeaderLogo(logoUrl)
            result.success("Logo set Successfully")
        }else if (call.method == "setHomeDashboardContent"){
             val submitButtonText = call.argument<String>("submitButtonText") ?: ""
              val ticketDescription = call.argument<String>("ticketDescription") ?: ""
               val ticketTitle = call.argument<String>("ticketTitle") ?: ""
                val kbDescription = call.argument<String>("kbDescription") ?: ""
                 val kbTitle = call.argument<String>("kbTitle") ?: ""
                  val headerDescription = call.argument<String>("headerDescription") ?: ""
                   val headerName = call.argument<String>("headerName") ?: ""

            BoldDeskSDKHome.setHomeDashboardContent(
                 headerName,
                 headerDescription,
                 kbTitle,
                 kbDescription,
                 ticketTitle,
                 ticketDescription,
                 submitButtonText,  
            )
            result.success("Content set Successfully")
        } else if( call.method == "isLoggedIn"){
            val value = BoldDeskSupportSDK.isLoggedIn(context)
            result.success(value)
        } else if (call.method == "clearallLocalData"){
            BoldDeskSupportSDK.clearAllLocalData(context)
            result.success("Cleared all local data Successfully")
        } else if (call.method == "isFromMobileSDK"){
           val messageData = call.argument<Map<String, String>>("userInfo") ?: emptyMap()
            val value = BoldDeskSupportSDK.isFromMobileSDK(messageData)
            result.success(value)
        } else if (call.method == "openArticleDetailsPage") {
            val articleId = call.argument<Int>("articleId") ?: 0
            val articleSlugTitle = call.argument<String>("articleSlugTitle") ?: ""
            BoldDeskSupportSDK.openArticleDetailsPage(context, articleId, articleSlugTitle)
            result.success("opened Article View Successfully")
        } else {
            result.notImplemented()
        }
    }

    private fun getFontResourceId(context: Context, fontName: String): Int {
        // This is the magic: It looks for a resource of type "font" with the given name
        // in the application's package.
        val resourceId = context.resources.getIdentifier(fontName, "font", context.packageName)

        if (resourceId == 0) {
            // This error is critical for debugging!
            throw Exception("Font resource '$fontName' not found. " +
                "Ensure it's in pubspec.yaml and the filename is correct.")
        }
        return resourceId
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun resolveIcon(context: Context, iconName: String): Int {
        // Try to resolve from drawable resources
        val resId = context.resources.getIdentifier(
            iconName,
            "drawable",
            context.packageName
        )

        // Return resolved ID or default launcher icon
        return if (resId != 0) resId else android.R.drawable.ic_dialog_info
    }
}

File: ios/Classes/BolddeskSupportSdkPlugin.swift
import BoldDeskSupportSDK
import Flutter
import UIKit

public class BolddeskSupportSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "bd_support_sdk", binaryMessenger: registrar.messenger())
    let instance = BolddeskSupportSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "showHomeDashboard":
      BDSupportSDK.showHomeDashboard()
      result(nil)
    case "showCreateTicket":
      BDSupportSDK.showCreateTicket()
      result(nil)
    case "showKB":
      BDSupportSDK.showKB()
      result(nil)
    case "openRecentTickets":
      BDSupportSDK.showRecentTicket()
      result(nil)
    case "setLoggingEnabled":
      BDSupportSDK.enableLogging()
      result(nil)
    case "logout":
      BDSupportSDK.logout()
      result(nil)
    case "applyCustomFontFamilyInIOS":
      guard let args = call.arguments as? [String: Any],
        let fontFamily = args["fontFamily"] as? String
      else {
        return
      }
      BDPortalConfiguration.customFontName = fontFamily
      result(nil)
    case "setSystemFontSize":
      guard let args = call.arguments as? [String: Any],
        let enable = args["enable"] as? Bool
      else {
        return
      }
      BDSupportSDK.applySystemFontSize = enable
      result(nil)

    case "applyTheme":
      guard let args = call.arguments as? [String: Any],
        let accentColor = args["accentColor"] as? String,
        let primaryColor = args["primaryColor"] as? String
      else {
        return
      }
      BDSupportSDK.applyTheme(
        accentColor: accentColor,
        primaryColor: primaryColor
      )
      result(nil)
    case "setPreferredTheme":
      guard let args = call.arguments as? [String: Any],
        let themeString = args["theme"] as? String
      else {
        result(["success": false, "message": "Missing theme"])
        return
      }

      let preferredTheme: SDKTheme
      switch themeString.lowercased() {
      case "light":
        preferredTheme = .light
      case "dark":
        preferredTheme = .dark
      default:
        preferredTheme = .system
      }

      BDSupportSDK.setPreferredTheme(preferredTheme)
      result(nil)
    case "initialize":
      guard let args = call.arguments as? [String: Any],
        let appId = args["appId"] as? String,
        let brandUrl = args["brandUrl"] as? String
      else {
        result(["success": false, "message": "Missing arguments"])
        return
      }

      BDSupportSDK.initialize(
        appId: appId,
        brandURl: brandUrl,
        { success in
          result(["success": true, "message": success ?? "Initialized successfully"])
        },
        { error in
          result(["success": false, "message": error ?? "Initialization failed"])
        }
      )

    case "loginWithJWTToken":
      guard let args = call.arguments as? [String: Any],
        let jwtToken = args["jwtToken"] as? String
      else {
        result(["success": false, "message": "Missing arguments"])
        return
      }

      BDSupportSDK.loginWithJWTToken(
        jwtToken: jwtToken ?? "",
        { success in
          result(["success": true, "message": success ?? "Login successfully"])
        },
        { error in
          result(["success": false, "message": error ?? "Login Failed"])
        }
      )
    case "handleNotification":
      if let args = call.arguments as? [String: Any] {
        BDSupportSDK.processRemoteNotification(userInfo: args)
        result(nil)
      } else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENTS", message: "Expected notification data", details: nil))
      }
    case "setFCMRegistrationToken":
      guard let args = call.arguments as? [String: Any],
        let token = args["token"] as? String
      else {
        return
      }
      BDSupportSDK.enablePushNotification(fcmToken: token)
      result(nil)
    case "setHeaderLogo":
      guard let args = call.arguments as? [String: Any],
        let logoURL = args["logoURL"] as? String
      else {
        return
      }
      BDSDKHome.setHeaderLogo(logoURL: logoURL)
      result(nil)
    case "clearallLocalData":
      BDSupportSDK.clearallLocalData()
      result(nil)
    case "isLoggedIn":
      let isLoggedIn = BDSupportSDK.isLoggedIn()
      result(isLoggedIn)
    case "isFromMobileSDK":
      if let userInfo = call.arguments as? [AnyHashable: Any] {
       let isFromSDK = BDSupportSDK.isFromMobileSDK(userInfo: userInfo)
            result(isFromSDK)
       } 
       else {
            result(false) // return false if no arguments
       }
    case "setHomeDashboardContent":
      guard let args = call.arguments as? [String: Any],
        let headerName = args["headerName"] as? String,
        let headerDescription = args["headerDescription"] as? String,
        let kbTitle = args["kbTitle"] as? String,
        let kbDescription = args["kbDescription"] as? String,
        let ticketTitle = args["ticketTitle"] as? String,
        let ticketDescription = args["ticketDescription"] as? String,
        let submitButtonText = args["submitButtonText"] as? String
      else {
        return
      }
      BDSDKHome.setHomeDashboardContent(
        headerName: headerName, headerDescription: headerDescription, kbTitle: kbTitle,
        kbDescription: kbDescription, ticketTitle: ticketTitle, ticketDescription: ticketDescription,
        submitButtonText: submitButtonText)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
